kong/tls/plugins/sni_filter.lua

```diff
@@ -0,0 +1,263 @@
+_M.POSTFIX_SNIS_PSEUDO_INDEX = POSTFIX_SNIS_PSEUDO_INDEX
+local TTL_FOREVER = { ttl = 0 }
+
+local ca_cert_cache_opts = {
+  l1_serializer = function(ca)
+    local x509, err = openssl_x509.new(ca.cert, "PEM")
+    if err then
+      return nil, err
+    end
+
+    return x509
+  end
+}
+
+
+-- make the table out side of function to reuse table
+local key = new_tab(1, 0)
+
+local function load_ca(ca_id)
+  kong.log.debug("cache miss for CA Cert")
+
+  key.id = ca_id
+  local ca, err = kong.db.ca_certificates:select(key)
+  if not ca then
+    if err then
+      return nil, err
+    end
+
+    return nil, "CA Certificate '" .. tostring(ca_id) .. "' does not exist"
+  end
+
+  return ca
+end
+
+local function merge_ca_ids(sni, ca_ids)
+  sni.ca_ids = sni.ca_ids or {}
+  local sni_ca_ids = sni.ca_ids
+
+  for _, ca_id in ipairs(ca_ids) do
+    if not sni_ca_ids[ca_id] then
+      sni_ca_ids[ca_id] = true
+    end
+  end
+end
+
+local function ca_cert_cache_key(ca_id)
+  return "mtls:cacert:" .. ca_id
+end
+
+local function load_routes_from_db(db, route_id, options)
+  kong.log.debug("cache miss for route id: " .. route_id.id)
+  local routes, err = db.routes:select(route_id, options)
+  if routes == nil then
+    -- the third value means "do not cache"
+    return nil, err, -1
+  end
+
+  return routes
+end
+
+
+local function build_snis_for_route(route, snis, send_ca_dn, ca_ids)
+  -- every route should have SNI or ask cert on all requests
+  if not route.snis or #route.snis == 0 then
+    snis["*"] = snis["*"] or {}
+
+    if send_ca_dn then
+      merge_ca_ids(snis["*"], ca_ids)
+    end
+
+  else
+    for _, sni in ipairs(route.snis) do
+      local sni_t
+      local idx = sni:find("*", 1, true)
+
+      if idx == 1 then
+        -- store snis with the leftmost wildcard in a subtable
+        snis[POSTFIX_SNIS_PSEUDO_INDEX] = snis[POSTFIX_SNIS_PSEUDO_INDEX] or {}
+        local postfix_snis = snis[POSTFIX_SNIS_PSEUDO_INDEX]
+        postfix_snis[sni] = postfix_snis[sni] or { value = sni:sub(2) }
+        sni_t = postfix_snis[sni]
+        kong.log.debug("add a postfix sni ", sni)
+
+      elseif idx == #sni then
+        -- store snis with the rightmost wildcard in a subtable
+        snis[PREFIX_SNIS_PSEUDO_INDEX] = snis[PREFIX_SNIS_PSEUDO_INDEX] or {}
+        local prefix_snis = snis[PREFIX_SNIS_PSEUDO_INDEX]
+        prefix_snis[sni] = prefix_snis[sni] or { value = sni:sub(1, -2) }
+        sni_t = prefix_snis[sni]
+        kong.log.debug("add a prefix sni ", sni)
+
+      else
+        snis[sni] = snis[sni] or {}
+        sni_t = snis[sni]
+        kong.log.debug("add a plain sni ", sni)
+      end
+
+      if send_ca_dn then
+        merge_ca_ids(sni_t, ca_ids)
+      end
+    end
+  end
+end
+
+
+local function get_snis_for_plugin(db, plugin, snis, options)
+  -- plugin applied on service
+  local service_pk = plugin.service
+  local send_ca_dn = plugin.config.send_ca_dn
+  local ca_ids = plugin.config.ca_certificates
```

Comment:
This is actually related to the mtls-auth plugin, and we may need to refactor here as well.